File: features_engine/src/regime/regime_filter.py
"""
Latent regime posterior P(Z_t = z | F_t) per math model Section 5.

v0: log-linear emission scores + softmax (not full HMM forward-backward).
Uses only F_t at time t; prior posterior is blended for temporal smoothing.
"""
from __future__ import annotations

import math
from typing import Dict

import numpy as np

REGIME_LABELS = (
    "normal",
    "event_shock",
    "liquidity_vacuum",
    "stop_cascade",
    "prop_flatten",
    "book_rebuild",
    "chop",
    "trend_continuation",
    "spread_stress",
)


def _softmax(logits: Dict[str, float]) -> Dict[str, float]:
    keys = list(logits.keys())
    vals = np.array([logits[k] for k in keys], dtype=np.float64)
    vals -= vals.max()
    ex = np.exp(vals)
    ex /= ex.sum()
    return {k: float(v) for k, v in zip(keys, ex)}


class RegimeFilter:
    """Estimates P(Z_t | F_t) from observable features at time t only."""

    def __init__(self, temperature: float = 1.0):
        self.temperature = temperature
        self._prev_posterior: Dict[str, float] = {
            z: 1.0 / len(REGIME_LABELS) for z in REGIME_LABELS
        }

    def update(
        self, features: Dict[str, float], event_context: str = "NORMAL"
    ) -> Dict[str, float]:
        spread_stress = features.get("spread_stress", 1.0)
        cancel_ratio = features.get("cancel_to_add_ratio", 1.0)
        agg_imb = features.get("aggressor_volume_imbalance", 0.0)
        slope_change = features.get("book_slope_change", 0.0)
        vacuum = features.get("liquidity_vacuum_score", 0.0)
        near_cancel = features.get("near_touch_cancel_pressure", 0.0)

        logits: Dict[str, float] = {z: 0.0 for z in REGIME_LABELS}

        logits["normal"] = -abs(agg_imb) - max(0.0, spread_stress - 1.0)
        logits["spread_stress"] = max(0.0, spread_stress - 1.2) * 2.0
        logits["liquidity_vacuum"] = vacuum * 3.0 + max(0.0, spread_stress - 1.0)
        logits["stop_cascade"] = abs(agg_imb) * 2.0 + max(0.0, cancel_ratio - 1.5)
        logits["event_shock"] = max(0.0, spread_stress - 1.5) + abs(slope_change) * 2.0
        logits["book_rebuild"] = (
            max(0.0, -slope_change * agg_imb) if slope_change != 0 else 0.0
        )
        logits["trend_continuation"] = abs(agg_imb) * 1.5 - abs(slope_change)
        logits["chop"] = -abs(agg_imb) + abs(slope_change) * 0.5
        logits["prop_flatten"] = 0.0

        if event_context in (
            "CPI_TIGHT",
            "NFP_TIGHT",
            "FOMC_STATEMENT_TIGHT",
            "FOMC_PRESS_TIGHT",
            "CASH_EQUITY_OPEN",
        ):
            logits["event_shock"] += 2.0
            logits["spread_stress"] += 1.0
        if event_context in (
            "PROP_FLATTEN_TOPSTEP",
            "TPT_FLATTEN",
            "APEX_FLATTEN",
            "FRIDAY_CLOSE",
        ):
            logits["prop_flatten"] += 3.0
            logits["normal"] -= 1.0
        if event_context == "NEWS_RESTRICTION":
            logits["prop_flatten"] += 1.5
        if near_cancel > 0.5:
            logits["event_shock"] += 0.5
            logits["liquidity_vacuum"] += 0.5

        for z in REGIME_LABELS:
            logits[z] += 0.15 * math.log(self._prev_posterior.get(z, 1e-9) + 1e-12)

        posterior = _softmax({z: logits[z] / self.temperature for z in REGIME_LABELS})
        self._prev_posterior = posterior
        return posterior

    @staticmethod
    def argmax(posterior: Dict[str, float]) -> str:
        return max(posterior, key=posterior.get)

    def volatility_state(self, features: Dict[str, float]) -> str:
        return "HIGH" if features.get("spread_stress", 1.0) > 1.5 else "NORMAL"

    def liquidity_state(self, features: Dict[str, float]) -> str:
        vacuum = features.get("liquidity_vacuum_score", 0.0)
        depth10 = features.get("top_10_depth_bid", 0.0) + features.get(
            "top_10_depth_ask", 0.0
        )
        if vacuum > 0.01 or depth10 < 50:
            return "THIN"
        return "NORMAL"

File: features_engine/cpp/src/regime_filter.cpp
#include "regime_filter.hpp"

#include "feature_index.hpp"

#include <algorithm>
#include <cmath>
#include <string>

namespace hft {

namespace {

constexpr size_t kRegimeCount = 9;

double clamp_nonneg(double x) { return x > 0.0 ? x : 0.0; }

void apply_event_context_boosts(double* logits, const std::string& event_context) {
    static const char* kEventShock[] = {
        "CPI_TIGHT", "NFP_TIGHT", "FOMC_STATEMENT_TIGHT", "FOMC_PRESS_TIGHT", "CASH_EQUITY_OPEN"};
    static const char* kPropFlatten[] = {
        "PROP_FLATTEN_TOPSTEP", "TPT_FLATTEN", "APEX_FLATTEN", "FRIDAY_CLOSE"};

    for (const char* tag : kEventShock) {
        if (event_context == tag) {
            logits[1] += 2.0;  // event_shock
            logits[8] += 1.0;  // spread_stress
            break;
        }
    }
    for (const char* tag : kPropFlatten) {
        if (event_context == tag) {
            logits[4] += 3.0;  // prop_flatten
            logits[0] -= 1.0;  // normal
            break;
        }
    }
    if (event_context == "NEWS_RESTRICTION") {
        logits[4] += 1.5;
    }
}

}  // namespace

RegimeFilterCpp::RegimeFilterCpp(double temperature) : temperature_(temperature) { reset(); }

void RegimeFilterCpp::reset() {
    const double uniform = 1.0 / static_cast<double>(kRegimeCount);
    prev_posterior_.fill(uniform);
}

void RegimeFilterCpp::update(std::array<double, 64>& vec, const std::string& event_context) {
    const double spread_stress = vec[static_cast<size_t>(FeatureIndex::SPREAD_STRESS)];
    const double cancel_ratio = vec[static_cast<size_t>(FeatureIndex::CANCEL_TO_ADD_RATIO)];
    const double agg_imb = vec[static_cast<size_t>(FeatureIndex::AGGRESSOR_VOLUME_IMBALANCE)];
    const double slope_change = vec[static_cast<size_t>(FeatureIndex::BOOK_SLOPE_CHANGE)];
    const double vacuum = vec[static_cast<size_t>(FeatureIndex::LIQUIDITY_VACUUM_SCORE)];
    const double near_cancel = vec[static_cast<size_t>(FeatureIndex::NEAR_TOUCH_CANCEL_PRESSURE)];

    double logits[kRegimeCount] = {0.0};

    // Order matches REGIME_LABELS in regime_filter.py
    logits[0] = -std::abs(agg_imb) - clamp_nonneg(spread_stress - 1.0);           // normal
    logits[1] = clamp_nonneg(spread_stress - 1.5) + std::abs(slope_change) * 2.0; // event_shock
    logits[2] = vacuum * 3.0 + clamp_nonneg(spread_stress - 1.0);                 // liquidity_vacuum
    logits[3] = std::abs(agg_imb) * 2.0 + clamp_nonneg(cancel_ratio - 1.5);       // stop_cascade
    logits[4] = 0.0;                                                              // prop_flatten
    logits[5] = slope_change != 0.0 ? clamp_nonneg(-slope_change * agg_imb) : 0.0;  // book_rebuild
    logits[6] = -std::abs(agg_imb) + std::abs(slope_change) * 0.5;                // chop
    logits[7] = std::abs(agg_imb) * 1.5 - std::abs(slope_change);                 // trend_continuation
    logits[8] = clamp_nonneg(spread_stress - 1.2) * 2.0;                          // spread_stress

    apply_event_context_boosts(logits, event_context);

    if (near_cancel > 0.5) {
        logits[1] += 0.5;
        logits[2] += 0.5;
    }

    for (size_t i = 0; i < kRegimeCount; ++i) {
        logits[i] += 0.15 * std::log(prev_posterior_[i] + 1e-12);
        logits[i] /= temperature_;
    }

    double max_logit = logits[0];
    for (size_t i = 1; i < kRegimeCount; ++i) max_logit = std::max(max_logit, logits[i]);

    double sum_exp = 0.0;
    double posterior[kRegimeCount];
    for (size_t i = 0; i < kRegimeCount; ++i) {
        posterior[i] = std::exp(logits[i] - max_logit);
        sum_exp += posterior[i];
    }
    for (size_t i = 0; i < kRegimeCount; ++i) {
        posterior[i] /= sum_exp;
        prev_posterior_[i] = posterior[i];
    }

    vec[static_cast<size_t>(FeatureIndex::REGIME_NORMAL)] = posterior[0];
    vec[static_cast<size_t>(FeatureIndex::REGIME_EVENT_SHOCK)] = posterior[1];
    vec[static_cast<size_t>(FeatureIndex::REGIME_LIQUIDITY_VACUUM)] = posterior[2];
    vec[static_cast<size_t>(FeatureIndex::REGIME_STOP_CASCADE)] = posterior[3];
    vec[static_cast<size_t>(FeatureIndex::REGIME_PROP_FLATTEN)] = posterior[4];
    vec[static_cast<size_t>(FeatureIndex::REGIME_BOOK_REBUILD)] = posterior[5];
    vec[static_cast<size_t>(FeatureIndex::REGIME_CHOP)] = posterior[6];
    vec[static_cast<size_t>(FeatureIndex::REGIME_TREND)] = posterior[7];
    vec[static_cast<size_t>(FeatureIndex::REGIME_SPREAD_STRESS)] = posterior[8];
}

}  // namespace hft
